dali/kernels/imgproc/warp_gpu.h

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_WARP_GPU_H_
+#define DALI_KERNELS_IMGPROC_WARP_GPU_H_
+
+#include "dali/core/common.h"
+#include "dali/core/geom/vec.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/imgproc/warp/warp_setup.cuh"
+#include "dali/kernels/imgproc/warp/warp_variable_size_impl.cuh"
+#include "dali/kernels/imgproc/warp/warp_uniform_size_impl.cuh"
+#include "dali/kernels/imgproc/warp/mapping_traits.h"
+
+namespace dali {
+namespace kernels {
+
+/// @remarks Assume HWC layout
+template <typename _Mapping, int ndim, typename _OutputType, typename _InputType,
+          typename _BorderValue>
+class WarpGPU : public warp::WarpSetup<ndim, _OutputType, _InputType> {
+ public:
+  static constexpr int spatial_ndim = ndim;
+  static constexpr int tensor_ndim = ndim + 1;
+
+  using Mapping = _Mapping;
+  using OutputType = _OutputType;
+  using InputType = _InputType;
+  using BorderValue = _BorderValue;
+  using MappingParams = warp::mapping_params_t<Mapping>;
+  static_assert(std::is_pod<MappingParams>::value, "Mapping parameters must be POD.");
+  static_assert(std::is_pod<BorderValue>::value, "BorderValue must be POD.");
+
+  using Base =  warp::WarpSetup<spatial_ndim, OutputType, InputType>;
+  using SampleDesc = typename Base::SampleDesc;
+  using BlockDesc = typename Base::BlockDesc;
+  static_assert(spatial_ndim == 2, "Not implemented for spatial_ndim != 2");
+
+  KernelRequirements Setup(KernelContext &context,
+                           const InListGPU<InputType, tensor_ndim> &in,
+                           const InTensorGPU<MappingParams, 1> &mapping,
+                           span<const TensorShape<spatial_ndim>> output_sizes,
```

Comment:
Yes, spatial component of size is provided externally. If the operator has ability to calculate it implicitly, then it would be the operator to provide the proper sizes.

Reply:
Ok